rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala

```diff
@@ -0,0 +1,850 @@
+        path: Path              // Return path
+    )
+
+    /**
+      * Create path from root to lastPrefix node
+      */
+    def initNodePath(
+        params: NodePathData
+    ): F[Either[NodePathData, Path]] =
+      params match {
+        case NodePathData(hash, nodePrefix, tempPrefix, path) =>
+          for {
+            nodeOpt <- getNodeDataFromStore(hash)
+            node <- nodeOpt.liftTo[F](
+                     new Exception(s"Export error: node with key ${hash.toHex} not found.")
+                   )
+            decodedNode = Codecs.decode(node)
+            r = if (tempPrefix.isEmpty)
+              (NodeData(nodePrefix, decodedNode, none) +: path).asRight // Happy end
+            else                                                        // Go dipper
+              decodedNode(byteToInt(tempPrefix.head)) match {
+                case NodePtr(ptrPrefix, ptr) =>
+                  val (prefixCommon, prefixRest, ptrPrefixRest) =
+                    commonPrefix(tempPrefix.tail, ptrPrefix)
+                  assert(
+                    ptrPrefixRest.isEmpty,
+                    s"Export error: node with prefix ${(nodePrefix ++ tempPrefix).toHex} not found."
+                  )
+                  NodePathData(
+                    ptr,
+                    (nodePrefix :+ tempPrefix.head) ++ prefixCommon,
+                    prefixRest,
+                    NodeData(nodePrefix, decodedNode, tempPrefix.head.some) +: path
+                  ).asLeft
+                case _ =>
+                  assert(
+                    assertion = false,
+                    s"Export error: node with prefix ${(nodePrefix ++ tempPrefix).toHex} not found."
+                  )
+                  Vector().asRight // Not found
+              }
+          } yield r
+      }
+
+    type LoopData = (
+        Path,       // Path of node from current to root
+        (Int, Int), // Skip & take counter
+        ExportData  // Result of export
+    )
+
+    /**
+      * Find next non-empty item.
+      *
+      * @param node Node to look for
+      * @param lastIdxOpt Last found index (if this node was not searched - [[None]])
+      * @return [[Some]](numItem, [[Item]]) if item found, [[None]] if non-empty item not found
+      */
+    @tailrec
+    def findNextNonEmptyItem(node: Node, lastIdxOpt: Option[Byte]): Option[(Byte, Item)] =
+      if (lastIdxOpt == 0xFF.toByte.some) none
+      else {
+        val curIdxInt = lastIdxOpt.map(byteToInt(_) + 1).getOrElse(0)
+        val curItem   = node(curIdxInt)
+        val curIdx    = curIdxInt.toByte
+        curItem match {
+          case EmptyItem => findNextNonEmptyItem(node, curIdx.some)
+          case Leaf(_, _) =>
+            if (settings.flagLPrefixes || settings.flagLValues) (curIdx, curItem).some
+            else findNextNonEmptyItem(node, curIdx.some)
+          case NodePtr(_, _) => (curIdx, curItem).some
+        }
+      }
```

Comment:
Consider this implementation. Maybe it is more clear?
```scala
def isLeaf(b: Item): Boolean = b match {
  case _: Leaf => true
  case _ => false
}
val item = node.find { item => isLeaf(item) && (settings.flagLPrefixes || settings.flagLValues) }
val index = item.map(node.indexOf(_))
```

Reply:
This is the same ques about what's better recursion or cycle. In your implementation missing a few things: using `lastIdxOpt`, returning Leafs and NodePtr (instead just Leaf). I understand that you show me an example, thanks! If @tgrospic agrees I will rewrite my code in the future.